implementation/src/main/java/io/smallrye/mutiny/operators/multi/builders/GeneratorBasedMulti.java

```diff
@@ -0,0 +1,135 @@
+package io.smallrye.mutiny.operators.multi.builders;
+
+import static io.smallrye.mutiny.helpers.ParameterValidation.nonNull;
+import static io.smallrye.mutiny.helpers.ParameterValidation.nonNullNpe;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Subscription;
+
+import io.smallrye.mutiny.groups.GeneratorEmitter;
+import io.smallrye.mutiny.helpers.Subscriptions;
+import io.smallrye.mutiny.operators.AbstractMulti;
+import io.smallrye.mutiny.subscription.MultiSubscriber;
+
+public class GeneratorBasedMulti<T, S> extends AbstractMulti<T> {
+
+    private final Supplier<S> initialStateSupplier;
+    private final BiFunction<S, GeneratorEmitter<? super T>, S> generator;
+
+    public GeneratorBasedMulti(Supplier<S> initialStateSupplier, BiFunction<S, GeneratorEmitter<? super T>, S> generator) {
+        this.initialStateSupplier = initialStateSupplier;
+        this.generator = generator;
+    }
+
+    @Override
+    public void subscribe(MultiSubscriber<? super T> subscriber) {
+        MultiSubscriber<? super T> downstream = nonNull(subscriber, "subscriber");
+        S initialState = initialStateSupplier.get();
```

Comment:
Should it handle exceptions here?

Reply:
My understanding is that since the subscription hasn't actually happened yet it is safe to let the exception go down the stack rather than subscribe + immediately fail.

Reply:
Well, if we have another operator downstream that won't necessary work (re-subscription, etc)

Reply:
So let me fix that

Reply:
IF the consumer uses `runSubscriptionOn`, the exception may not be catchable.

Reply:
Fixed, see 10e0067